File: best-time-to-buy-and-sell-stock.cpp
//https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(prices.size()==0) return 0;
        int min = prices[0];
        int max_profit = 0;
        for(int i=1 ; i<prices.size() ; i++){
            if(prices[i]<min) min=prices[i];
            else if(prices[i]-min>max_profit) max_profit=prices[i]-min;
        }
        return max_profit;
    }
};

File: 121/best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
    mn := 1 << 32
    max_profit := 0
    
    for _, val := range prices {
        mn = min(mn, val)
        max_profit = max(max_profit, val - mn)
    }
    
    return max_profit
}

func max(i, j int) int {
    if i >= j {
        return i
    }
    
    return j
}

func min(i, j int) int {
    if i < j {
        return i
    }
    
    return j
}

